### PR TITLE
Gh 595 add serviceinstancelistsupplier

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**

### CODE REVIEW

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactiveLoadBalancerAutoConfiguration.java [hunk 1]
```diff
@@ -39,6 +39,7 @@
 @ConditionalOnClass(WebClient.class)
 @ConditionalOnBean(LoadBalancerClient.class)
 @ConditionalOnMissingBean(ReactiveLoadBalancer.Factory.class)
+@Deprecated
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactiveLoadBalancerAutoConfiguration.java]: Is this related to this PR?

File: spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/RoundRobinLoadBalancer.java [hunk 2]
```diff
@@ -62,19 +111,30 @@ public RoundRobinLoadBalancer(String serviceId,
 	// src/main/java/netflix/ocelli/loadbalancer/RoundRobinLoadBalancer.java
 	public Mono<Response<ServiceInstance>> choose(Request request) {
 		// TODO: move supplier to Request?
+		// Temporary conditional logic till deprecated members are removed.
+		if (serviceInstanceListSupplierProvider != null) {
+			ServiceInstanceListSupplier supplier = serviceInstanceListSupplierProvider
+					.getIfAvailable();
+			return Objects.requireNonNull(supplier).get().next()
```
Comment (spencergibb) [inline spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/RoundRobinLoadBalancer.java]: I'd rather not have `Objects.requireNonNull()` here. If it's null, it just throws a NPE which isn't helpful. Add a default supplier to `getIfAvailable()` that just returns a noop one.

File: spring-cloud-loadbalancer/src/test/java/org/springframework/cloud/loadbalancer/core/LoadBalancerTests.java [hunk 3]
```diff
@@ -116,17 +116,17 @@ public void emptyHosts() {
 	@Test
 	public void staticConfigurationWorks() {
```
Comment (spencergibb) [inline spring-cloud-loadbalancer/src/test/java/org/springframework/cloud/loadbalancer/core/LoadBalancerTests.java]: Can we add a new test for the list supplier so we can keep the existing test until it goes away.

Comment (OlgaMaciaszek) [reply spring-cloud-loadbalancer/src/test/java/org/springframework/cloud/loadbalancer/core/LoadBalancerTests.java]: We could - I just wanted to remove deprecated elements wherever possible.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactiveLoadBalancerAutoConfiguration.java (hunk 1 above)
Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactiveLoadBalancerAutoConfiguration.java]: Not much, have realised it was missing.

Comment (OlgaMaciaszek) [reply spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/reactive/ReactiveLoadBalancerAutoConfiguration.java]: Will move it to a separate comit.

File: spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/CachingServiceInstanceListSupplier.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	private final Flux<List<ServiceInstance>> serviceInstances;
+
+	@SuppressWarnings("unchecked")
+	public CachingServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
+			CacheManager cacheManager) {
+		this.delegate = delegate;
+		this.serviceInstances = CacheFlux.lookup(key -> {
+			Cache cache = cacheManager.getCache(SERVICE_INSTANCE_CACHE_NAME); // TODO:
+			// configurable
```
Comment (ryanjbaxter) [inline spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/CachingServiceInstanceListSupplier.java]: Can we fix the formatting of this comment?

File: spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/RoundRobinLoadBalancer.java [hunk 5]
```diff
@@ -62,19 +111,30 @@ public RoundRobinLoadBalancer(String serviceId,
...
-				return new EmptyResponse();
-			}
-			// TODO: enforce order?
-			int pos = Math.abs(this.position.incrementAndGet());
-
-			ServiceInstance instance = instances.get(pos % instances.size());
-
-			return new DefaultResponse(instance);
-		});
+		return Objects.requireNonNull(supplier).get().collectList()
```
Comment (ryanjbaxter) [inline spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/RoundRobinLoadBalancer.java]: Same here?

Comment (spencergibb) [reply spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/RoundRobinLoadBalancer.java]: yes

File: spring-cloud-loadbalancer/src/test/java/org/springframework/cloud/loadbalancer/core/LoadBalancerTests.java (hunk 3 above)
Comment (OlgaMaciaszek) [reply spring-cloud-loadbalancer/src/test/java/org/springframework/cloud/loadbalancer/core/LoadBalancerTests.java]: Will add it back.
